mvn build issues

I'm not the best maven person, but I'm pretty sure something needs to be added to the README for connecting to the bcgov specific jar files references in the pom.xml file.  Thanks!

[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building did-auth-relying-party 0.1-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[WARNING] The POM for fi.trustnet:verifiable-credentials-java:jar:0.1-SNAPSHOT is missing, no dependency information available
[WARNING] The POM for decentralized-identity:uni-resolver-client:jar:0.1-SNAPSHOT is missing, no dependency information available
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 0.263 s
[INFO] Finished at: 2018-03-21T16:48:48-04:00
[INFO] Final Memory: 10M/309M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project did-auth-relying-party: Could not resolve dependencies for project ca.bcgov.didauth:did-auth-relying-party:war:0.1-SNAPSHOT: The following artifacts could not be resolved: fi.trustnet:verifiable-credentials-java:jar:0.1-SNAPSHOT, decentralized-identity:uni-resolver-client:jar:0.1-SNAPSHOT: Could not find artifact fi.trustnet:verifiable-credentials-java:jar:0.1-SNAPSHOT -> [Help 1]

--- a/README.md
+++ b/README.md
@@ -41,6 +41,13 @@ DID Auth relying party sample deployment: https://did-auth-relying-party.danubet
 
 ## Deployment (Local Development)
 
+First, build the following Maven dependencies:
+
+ * https://github.com/WebOfTrustInfo/ld-signatures-java
+ * https://github.com/TrustNetFI/verifiable-credentials-java
+ * https://github.com/WebOfTrustInfo/txref-conversion-java.git
+ * https://github.com/decentralized-identity/universal-resolver.git
+
 Build with Maven:
 
 	mvn clean install jetty:run

--- a/docs/scenarios.md
+++ b/docs/scenarios.md
@@ -62,7 +62,7 @@ This is a Verifiable Credential issued by an "Org Manager" that an "Org Delegate
 
 **Org Manager DID:** `did:sov:DavnUKB3kjn7VmVZXzEDL7`
 
-**Org Delegate DID:** `did:v1:test:nym:rZdPg5VF6SqrVuEYEHAuDaeikkA2D8QBLRJQRnhz3pI`
+**Org Delegate DID:** `did:v1:test:nym:UxYjr6F3hqwiF3yffplpcsV3pXSWSzVQ2396WT65e2E`
 
 	{
 		"@context" : [ "https://w3id.org/credentials/v1", "https://w3id.org/security/v1" ],
@@ -71,7 +71,7 @@ This is a Verifiable Credential issued by an "Org Manager" that an "Org Delegate
 		"issued" : "2018-03-15",
 		"claim" : {
 			"id" : "did:sov:1234567890",
-			"management" : "did:v1:test:nym:rZdPg5VF6SqrVuEYEHAuDaeikkA2D8QBLRJQRnhz3pI"
+			"management" : "did:v1:test:nym:UxYjr6F3hqwiF3yffplpcsV3pXSWSzVQ2396WT65e2E"
 		},
 		"signature" : {
 			"type" : "Ed25519Signature2018",

--- a/docs/scenarios.md
+++ b/docs/scenarios.md
@@ -7,19 +7,21 @@ This is simply a self-asserted Verifiable Credential by an identity owner to aut
 **Admin DID:** `did:btcr:xs95-wzv8-qqte-jz88`
 
 	{
-		"type": ["Credential", "DidAuthCredential"],
-		"issuer": "did:btcr:xs95-wzv8-qqte-jz88",
-		"issued": "2010-01-01",
-		"claim": {
-			"id": "did:btcr:xs95-wzv8-qqte-jz88",
-			"publicKey": "did:btcr:xs95-wzv8-qqte-jz88#keys-1"
+		"@context" : [ "https://w3id.org/credentials/v1", "https://w3id.org/security/v1" ],
+		"type" : [ "Credential", "DidAuthCredential" ],
+		"issuer" : "did:btcr:xs95-wzv8-qqte-jz88",
+		"issued" : "2018-03-15",
+		"claim" : {
+			"id" : "did:btcr:xs95-wzv8-qqte-jz88",
+			"publicKey" : "did:btcr:xs95-wzv8-qqte-jz88#key-1"
 		},
-		"proof": {
-			"type": "Secp256k1Signature2018",
-			"creator": "did:btcr:xs95-wzv8-qqte-jz88#keys-1",
-			"created": "2018-01-01T21:19:10Z",
-			"nonce": "...",
-			"signatureValue": "..."
+		"signature" : {
+			"type" : "EcdsaKoblitzSignature2016",
+			"creator" : "did:btcr:xs95-wzv8-qqte-jz88#key-1",
+			"created" : "2018-03-15T00:00:00Z",
+			"domain" : null,
+			"nonce" : "54c83860-a79b-4523-88f9-ad420c669c12",
+			"signatureValue" : "MEUCIQC8JxrztioLbR5dfQwLBPnKMfE6RObSU9jpAJUr+YBxSQIgCbTbiabx8DMeKKgW1BfT+c+U9fy7DnhYjfH1xGFU8GY="
 		}
 	}
 
@@ -27,26 +29,28 @@ This is simply a self-asserted Verifiable Credential by an identity owner to aut
 
 This is a Verifiable Credential issued by VON that an "Org Manager" is authorized to represent an "Org" with capability "management".
 
-**TheOrgBook DID:** `did:sov:0000000000`
+**TheOrgBook DID:** `did:sov:MDBKSD4Cm5EhhXWzTynube`
 
 **Org DID:** `did:sov:1234567890`
 
 **Org Manager DID:** `did:sov:DavnUKB3kjn7VmVZXzEDL7`
 
 	{
-		"type": ["Credential"],
-		"issuer": "did:sov:0000000000",
-		"issued": "2010-01-01",
-		"claim": {
-			"id": "did:sov:1234567890",
-			"management": "did:sov:DavnUKB3kjn7VmVZXzEDL7"
+		"@context" : [ "https://w3id.org/credentials/v1", "https://w3id.org/security/v1" ],
+		"type" : [ "Credential" ],
+		"issuer" : "did:sov:MDBKSD4Cm5EhhXWzTynube",
+		"issued" : "2018-03-15",
+		"claim" : {
+			"id" : "did:sov:1234567890",
+			"management" : "did:sov:DavnUKB3kjn7VmVZXzEDL7"
 		},
-		"proof": {
-			"type": "Ed25519Signature2018",
-			"created": "2018-01-01T21:19:10Z",
-			"creator": "did:sov:0000000000#keys-1",
-			"nonce": "...",
-			"signatureValue": "..."
+		"signature" : {
+			"type" : "Ed25519Signature2018",
+			"creator" : "did:sov:MDBKSD4Cm5EhhXWzTynube#key-1",
+			"created" : "2018-03-15T00:00:00Z",
+			"domain" : null,
+			"nonce" : "d1458980-d63b-4c3c-aad9-f1d182f5b399",
+			"signatureValue" : "CGG+Isa9QtpqS9NuKTATBipY2cb50fFlCKHyLPV+Z136m304IqHjWnP9M0QM6i4ilYuUcS2ibOJtjWPHfmk7AQ=="
 		}
 	}
 
@@ -61,18 +65,20 @@ This is a Verifiable Credential issued by an "Org Manager" that an "Org Delegate
 **Org Delegate DID:** `did:v1:test:nym:rZdPg5VF6SqrVuEYEHAuDaeikkA2D8QBLRJQRnhz3pI`
 
 	{
-		"type": ["Credential"],
-		"issuer": "did:sov:DavnUKB3kjn7VmVZXzEDL7",
-		"issued": "2010-01-01",
-		"claim": {
-			"id": "did:sov:1234567890",
-			"management": "did:v1:test:nym:rZdPg5VF6SqrVuEYEHAuDaeikkA2D8QBLRJQRnhz3pI"
+		"@context" : [ "https://w3id.org/credentials/v1", "https://w3id.org/security/v1" ],
+		"type" : [ "Credential" ],
+		"issuer" : "did:sov:DavnUKB3kjn7VmVZXzEDL7",
+		"issued" : "2018-03-15",
+		"claim" : {
+			"id" : "did:sov:1234567890",
+			"management" : "did:v1:test:nym:rZdPg5VF6SqrVuEYEHAuDaeikkA2D8QBLRJQRnhz3pI"
 		},
-		"proof": {
-			"type": "Ed25519Signature2018",
-			"created": "2018-01-01T21:19:10Z",
-			"creator": "did:sov:DavnUKB3kjn7VmVZXzEDL7#keys-1",
-			"nonce": "c0ae1c8e-c7e7-469f-b252-86e6a0e7387e",
-			"signatureValue": "..."
+		"signature" : {
+			"type" : "Ed25519Signature2018",
+			"creator" : "did:sov:DavnUKB3kjn7VmVZXzEDL7#key-1",
+			"created" : "2018-03-15T00:00:00Z",
+			"domain" : null,
+			"nonce" : "17eb96ae-ccef-4500-bb5b-01e92a93abe7",
+			"signatureValue" : "YuYSql92rC1PloE3TNSD/4LJ15I3zbbMtYB4D6BSWdnuvOJBX1WobbDC0/tAcuJ0xWkL47DpOibJXFsKIjlsDw=="
 		}
 	}

--- a/pom.xml
+++ b/pom.xml
@@ -138,6 +138,18 @@
 			<version>0.1-SNAPSHOT</version>
 			<scope>compile</scope>
 		</dependency>
+		<dependency>
+			<groupId>org.bitcoinj</groupId>
+			<artifactId>bitcoinj-core</artifactId>
+			<version>0.14.5</version>
+			<scope>compile</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.abstractj.kalium</groupId>
+			<artifactId>kalium</artifactId>
+			<version>0.7.0</version>
+			<scope>compile</scope>
+		</dependency>
 
 	</dependencies>
 

--- a/src/main/java/ca/bcgov/didauth/rp/CallbackServlet.java
+++ b/src/main/java/ca/bcgov/didauth/rp/CallbackServlet.java
@@ -4,6 +4,7 @@ import java.io.IOException;
 import java.security.GeneralSecurityException;
 import java.util.ArrayList;
 import java.util.LinkedHashMap;
+import java.util.LinkedList;
 
 import javax.servlet.ServletException;
 import javax.servlet.http.HttpServlet;
@@ -31,6 +32,7 @@ public class CallbackServlet extends HttpServlet {
 		ArrayList<LinkedHashMap<String, Object>> jsonArray = (ArrayList<LinkedHashMap<String, Object>>) JsonUtils.fromReader(request.getReader());
 		if (log.isInfoEnabled()) log.info("JSON:" + jsonArray);
 
+		String authenticatedDid = null;
 		ArrayList<VerifiableCredential> vcs = new ArrayList<VerifiableCredential> ();
 		ArrayList<VerifiableCredential> authorizations = new ArrayList<VerifiableCredential> ();
 
@@ -61,21 +63,41 @@ public class CallbackServlet extends HttpServlet {
 
 			// DID Auth Verifiable Credential?
 
-			String authenticatedDid = Interpretation.interpretDidAuthCredential(vc);
+			if (authenticatedDid == null) {
 
-			if (authenticatedDid != null) {
+				authenticatedDid = Interpretation.interpretDidAuthCredential(vc);
 
-				request.getSession().setAttribute("authenticatedDid", authenticatedDid);
-			} else {
+				if (authenticatedDid == null) {
 
-				VerifiableCredential authorization = vc;
-				authorizations.add(authorization);
+					VerifiableCredential authorization = vc;
+					authorizations.add(authorization);
+				}
 			}
 		}
 
-		// done
+		// set session attributes
 
+		if (authenticatedDid != null) request.getSession().setAttribute("authenticatedDid", authenticatedDid);
 		request.getSession().setAttribute("vcs", vcs);
 		request.getSession().setAttribute("authorizations", authorizations);
+
+		// write response
+
+		LinkedHashMap<String, Object> jsonLdObject = new LinkedHashMap<String, Object> ();
+
+		if (authenticatedDid != null) jsonLdObject.put("authenticatedDid", authenticatedDid);
+
+		LinkedList<Object> jsonVcs = new LinkedList<Object> ();
+		for (VerifiableCredential vc : vcs) jsonVcs.add(vc.getJsonLdObject());
+		jsonLdObject.put("vcs", jsonVcs);
+
+		LinkedList<Object> jsonAuthorizations = new LinkedList<Object> ();
+		for (VerifiableCredential authorization : authorizations) jsonAuthorizations.add(authorization.getJsonLdObject());
+		jsonLdObject.put("authorizations", jsonAuthorizations);
+
+		response.setContentType("application/json");
+		response.getWriter().println(JsonUtils.toString(jsonLdObject));
+		response.getWriter().flush();
+		response.getWriter().close();
 	}
 }

--- a/src/main/java/ca/bcgov/didauth/rp/CallbackServlet.java
+++ b/src/main/java/ca/bcgov/didauth/rp/CallbackServlet.java
@@ -1,6 +1,7 @@
 package ca.bcgov.didauth.rp;
 
 import java.io.IOException;
+import java.security.GeneralSecurityException;
 import java.util.ArrayList;
 import java.util.LinkedHashMap;
 
@@ -12,9 +13,11 @@ import javax.servlet.http.HttpServletResponse;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import com.github.jsonldjava.core.JsonLdError;
 import com.github.jsonldjava.utils.JsonUtils;
 
 import fi.trustnet.verifiablecredentials.VerifiableCredential;
+import uniresolver.ResolutionException;
 
 public class CallbackServlet extends HttpServlet {
 
@@ -31,13 +34,32 @@ public class CallbackServlet extends HttpServlet {
 		ArrayList<VerifiableCredential> vcs = new ArrayList<VerifiableCredential> ();
 		ArrayList<VerifiableCredential> authorizations = new ArrayList<VerifiableCredential> ();
 
-		// interpret Verifiable Credentials
+		// interpret and verify Verifiable Credentials
 
 		for (LinkedHashMap<String, Object> jsonLdObject : jsonArray) {
 
 			VerifiableCredential vc = Interpretation.interpretCredential(VerifiableCredential.fromJsonLdObject(jsonLdObject));
 			if (vc == null) continue;
+
+			// verify
+
+			try {
+
+				boolean verify = Verification.verify(vc);
+				if (! verify) throw new GeneralSecurityException("Invalid signature.");
+			} catch (ResolutionException | GeneralSecurityException | JsonLdError ex) {
+
+				if (log.isWarnEnabled()) log.warn("Unable to verify LD signature: " + ex.getMessage(), ex);
+				response.sendError(HttpServletResponse.SC_FORBIDDEN, "Unable to verify LD signature: " + ex.getMessage());
+				response.getWriter().close();
+				return;
+			}
+
+			// add Verifiable Credential
+
 			vcs.add(vc);
+
+			// DID Auth Verifiable Credential?
 
 			String authenticatedDid = Interpretation.interpretDidAuthCredential(vc);
 

--- a/src/main/java/ca/bcgov/didauth/rp/Interpretation.java
+++ b/src/main/java/ca/bcgov/didauth/rp/Interpretation.java
@@ -4,7 +4,7 @@ import fi.trustnet.verifiablecredentials.VerifiableCredential;
 
 public class Interpretation {
 
-	static String interpretDidAuthCredential(VerifiableCredential vc) {
+	public static String interpretDidAuthCredential(VerifiableCredential vc) {
 
 		if (vc.getIssuer() == null || vc.getSubject() == null) return null;
 
@@ -15,7 +15,7 @@ public class Interpretation {
 		return vc.getSubject();
 	}
 
-	static VerifiableCredential interpretCredential(VerifiableCredential vc) {
+	public static VerifiableCredential interpretCredential(VerifiableCredential vc) {
 
 		if (vc.getIssuer() == null || vc.getSubject() == null) return null;
 

--- a/src/main/java/ca/bcgov/didauth/rp/Verification.java
+++ b/src/main/java/ca/bcgov/didauth/rp/Verification.java
@@ -1,0 +1,83 @@
+package ca.bcgov.didauth.rp;
+
+import java.net.URI;
+import java.security.GeneralSecurityException;
+import java.security.interfaces.RSAPublicKey;
+
+import org.bitcoinj.core.ECKey;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.jsonldjava.core.JsonLdError;
+
+import ca.bcgov.didauth.rp.verificationkeys.Ed25519VerificationKey2018Loader;
+import ca.bcgov.didauth.rp.verificationkeys.RsaVerificationKey2018Loader;
+import ca.bcgov.didauth.rp.verificationkeys.Secp256k1VerificationKey2018Loader;
+import fi.trustnet.verifiablecredentials.VerifiableCredential;
+import info.weboftrust.ldsignatures.LdSignature;
+import info.weboftrust.ldsignatures.suites.SignatureSuites;
+import info.weboftrust.ldsignatures.validator.EcdsaKoblitzSignature2016LdValidator;
+import info.weboftrust.ldsignatures.validator.Ed25519Signature2018LdValidator;
+import info.weboftrust.ldsignatures.validator.LdValidator;
+import info.weboftrust.ldsignatures.validator.RsaSignature2017LdValidator;
+import uniresolver.ResolutionException;
+import uniresolver.client.ClientUniResolver;
+import uniresolver.did.DIDDocument;
+import uniresolver.did.PublicKey;
+import uniresolver.result.ResolutionResult;
+
+public class Verification {
+
+	private static Logger log = LoggerFactory.getLogger(Verification.class);
+
+	public static boolean verify(VerifiableCredential vc) throws ResolutionException, GeneralSecurityException, JsonLdError {
+
+		// discover issuer's verification key
+
+		ClientUniResolver clientUniResolver = new ClientUniResolver();
+		clientUniResolver.setResolveUri("https://uniresolver.io/1.0/identifiers/");
+
+		URI issuer = vc.getIssuer();
+		if (log.isDebugEnabled()) log.debug("Issuer DID: " + issuer.toString());
+
+		LdSignature ldSignature = vc.getLdSignature();
+		if (ldSignature == null) {
+
+			if (log.isWarnEnabled()) log.warn("No LD signature found on Verifiable Credential.");
+			return false;
+		}
+
+		String signatureType = ldSignature.getType();
+		if (log.isDebugEnabled()) log.debug("LD signature type: " + signatureType);
+
+		ResolutionResult resolutionResult = clientUniResolver.resolve(issuer.toString());
+		DIDDocument didDocument = resolutionResult.getDidDocument();
+
+		LdValidator<?> ldValidator = null;
+		Object verifyingKey = null;
+
+		if (SignatureSuites.SIGNATURE_SUITE_ED25519SIGNATURE2018.getTerm().equals(signatureType)) {
+
+			for (PublicKey didPublicKey : didDocument.getPublicKeys()) if (verifyingKey == null) verifyingKey = Ed25519VerificationKey2018Loader.get().verificationKey(didPublicKey);
+			ldValidator = new Ed25519Signature2018LdValidator((byte[]) verifyingKey);
+		} else if (SignatureSuites.SIGNATURE_SUITE_ECDSAKOBLITZSIGNATURE2016.getTerm().equals(signatureType)) {
+
+			for (PublicKey didPublicKey : didDocument.getPublicKeys()) if (verifyingKey == null) verifyingKey = Secp256k1VerificationKey2018Loader.get().verificationKey(didPublicKey);
+			ldValidator = new EcdsaKoblitzSignature2016LdValidator((ECKey) verifyingKey);
+		} else if (SignatureSuites.SIGNATURE_SUITE_RSASIGNATURE2017.getTerm().equals(signatureType)) {
+
+			for (PublicKey didPublicKey : didDocument.getPublicKeys()) if (verifyingKey == null) verifyingKey = RsaVerificationKey2018Loader.get().verificationKey(didPublicKey);
+			ldValidator = new RsaSignature2017LdValidator((RSAPublicKey) verifyingKey);
+		}
+
+		if (log.isDebugEnabled()) log.debug("LD validator: " + ldValidator.getClass().getSimpleName());
+		if (log.isDebugEnabled()) log.debug("Verifying key: " + verifyingKey.getClass().getSimpleName());
+
+		if (ldValidator == null || verifyingKey == null) return false;
+
+		boolean verified = ldValidator.validate(vc.getJsonLdObject());
+		if (log.isDebugEnabled()) log.debug("Verified: " + verified);
+
+		return verified;
+	}
+}

--- a/src/main/java/ca/bcgov/didauth/rp/Verification.java
+++ b/src/main/java/ca/bcgov/didauth/rp/Verification.java
@@ -3,6 +3,8 @@ package ca.bcgov.didauth.rp;
 import java.net.URI;
 import java.security.GeneralSecurityException;
 import java.security.interfaces.RSAPublicKey;
+import java.util.ArrayList;
+import java.util.List;
 
 import org.bitcoinj.core.ECKey;
 import org.slf4j.Logger;
@@ -22,6 +24,7 @@ import info.weboftrust.ldsignatures.validator.LdValidator;
 import info.weboftrust.ldsignatures.validator.RsaSignature2017LdValidator;
 import uniresolver.ResolutionException;
 import uniresolver.client.ClientUniResolver;
+import uniresolver.did.Authentication;
 import uniresolver.did.DIDDocument;
 import uniresolver.did.PublicKey;
 import uniresolver.result.ResolutionResult;
@@ -39,12 +42,15 @@ public class Verification {
 
 		URI issuer = vc.getIssuer();
 		if (log.isDebugEnabled()) log.debug("Issuer DID: " + issuer.toString());
+		if (issuer == null) {
+
+			throw new GeneralSecurityException("No issuer found on Verifiable Credential.");
+		}
 
 		LdSignature ldSignature = vc.getLdSignature();
 		if (ldSignature == null) {
 
-			if (log.isWarnEnabled()) log.warn("No LD signature found on Verifiable Credential.");
-			return false;
+			throw new GeneralSecurityException("No LD signature found on Verifiable Credential.");
 		}
 
 		String signatureType = ldSignature.getType();
@@ -52,21 +58,33 @@ public class Verification {
 
 		ResolutionResult resolutionResult = clientUniResolver.resolve(issuer.toString());
 		DIDDocument didDocument = resolutionResult.getDidDocument();
+		if (didDocument == null) throw new ResolutionException("No resolution result for DID " + issuer.toString());
+
+		List<PublicKey> didPublicKeys = new ArrayList<PublicKey> ();
+		if (didDocument.getPublicKeys() != null) didPublicKeys.addAll(didDocument.getPublicKeys());
+		if (didDocument.getAuthentications() != null) {
+
+			for (Authentication didAuthentication : didDocument.getAuthentications()) {
+
+				if (didAuthentication.getPublicKeys() != null) didPublicKeys.addAll(didAuthentication.getPublicKeys());
+			}
+		}
+		if (log.isDebugEnabled()) log.debug("DID public keys: " + didPublicKeys);
 
 		LdValidator<?> ldValidator = null;
 		Object verifyingKey = null;
 
 		if (SignatureSuites.SIGNATURE_SUITE_ED25519SIGNATURE2018.getTerm().equals(signatureType)) {
 
-			for (PublicKey didPublicKey : didDocument.getPublicKeys()) if (verifyingKey == null) verifyingKey = Ed25519VerificationKey2018Loader.get().verificationKey(didPublicKey);
+			for (PublicKey didPublicKey : didPublicKeys) if (verifyingKey == null) verifyingKey = Ed25519VerificationKey2018Loader.get().verificationKey(didPublicKey);
 			ldValidator = new Ed25519Signature2018LdValidator((byte[]) verifyingKey);
 		} else if (SignatureSuites.SIGNATURE_SUITE_ECDSAKOBLITZSIGNATURE2016.getTerm().equals(signatureType)) {
 
-			for (PublicKey didPublicKey : didDocument.getPublicKeys()) if (verifyingKey == null) verifyingKey = Secp256k1VerificationKey2018Loader.get().verificationKey(didPublicKey);
+			for (PublicKey didPublicKey : didPublicKeys) if (verifyingKey == null) verifyingKey = Secp256k1VerificationKey2018Loader.get().verificationKey(didPublicKey);
 			ldValidator = new EcdsaKoblitzSignature2016LdValidator((ECKey) verifyingKey);
 		} else if (SignatureSuites.SIGNATURE_SUITE_RSASIGNATURE2017.getTerm().equals(signatureType)) {
 
-			for (PublicKey didPublicKey : didDocument.getPublicKeys()) if (verifyingKey == null) verifyingKey = RsaVerificationKey2018Loader.get().verificationKey(didPublicKey);
+			for (PublicKey didPublicKey : didPublicKeys) if (verifyingKey == null) verifyingKey = RsaVerificationKey2018Loader.get().verificationKey(didPublicKey);
 			ldValidator = new RsaSignature2017LdValidator((RSAPublicKey) verifyingKey);
 		}
 

--- a/src/main/java/ca/bcgov/didauth/rp/verificationkeys/Ed25519VerificationKey2018Loader.java
+++ b/src/main/java/ca/bcgov/didauth/rp/verificationkeys/Ed25519VerificationKey2018Loader.java
@@ -1,0 +1,50 @@
+package ca.bcgov.didauth.rp.verificationkeys;
+
+import java.security.GeneralSecurityException;
+
+import org.abstractj.kalium.NaCl;
+import org.apache.commons.codec.binary.Base64;
+import org.bitcoinj.core.Base58;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.spongycastle.util.encoders.Hex;
+
+public class Ed25519VerificationKey2018Loader implements VerificationKeyLoader<byte[]> {
+
+	private static final String DID_PUBLIC_KEY_TYPE = "Ed25519VerificationKey2018";
+
+	private static Logger log = LoggerFactory.getLogger(Ed25519VerificationKey2018Loader.class);
+	private static final Ed25519VerificationKey2018Loader instance = new Ed25519VerificationKey2018Loader();
+
+	static {
+
+		NaCl.init();
+	}
+
+	public static Ed25519VerificationKey2018Loader get() {
+
+		return instance;
+	}
+
+	@Override
+	public byte[] verificationKey(uniresolver.did.PublicKey didPublicKey) throws GeneralSecurityException {
+
+		if (! didPublicKey.isType(DID_PUBLIC_KEY_TYPE)) {
+
+			if (log.isDebugEnabled()) log.debug("Public key " + didPublicKey.getId() + " is not of type " + DID_PUBLIC_KEY_TYPE);
+			return null;
+		}
+
+		String verificationKeyString;
+
+		verificationKeyString = didPublicKey.getPublicKeyBase64();
+		if (verificationKeyString != null) return Base64.decodeBase64(verificationKeyString);
+		verificationKeyString = didPublicKey.getPublicKeyBase58();
+		if (verificationKeyString != null) return Base58.decode(verificationKeyString);
+		verificationKeyString = didPublicKey.getPublicKeyHex();
+		if (verificationKeyString != null) return Hex.decode(verificationKeyString);
+
+		if (log.isDebugEnabled()) log.debug("Public key " + didPublicKey.getId() + " has no data.");
+		return null;
+	}
+}

--- a/src/main/java/ca/bcgov/didauth/rp/verificationkeys/RsaVerificationKey2018Loader.java
+++ b/src/main/java/ca/bcgov/didauth/rp/verificationkeys/RsaVerificationKey2018Loader.java
@@ -1,0 +1,64 @@
+package ca.bcgov.didauth.rp.verificationkeys;
+
+import java.security.GeneralSecurityException;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.interfaces.RSAPublicKey;
+import java.security.spec.X509EncodedKeySpec;
+
+import org.apache.commons.codec.binary.Base64;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RsaVerificationKey2018Loader implements VerificationKeyLoader<RSAPublicKey> {
+
+	private static final String DID_PUBLIC_KEY_TYPE = "RsaVerificationKey2018";
+
+	private static Logger log = LoggerFactory.getLogger(RsaVerificationKey2018Loader.class);
+	private static final RsaVerificationKey2018Loader instance = new RsaVerificationKey2018Loader();
+
+	private static KeyFactory keyFactory;
+
+	static {
+
+		try {
+
+			keyFactory = KeyFactory.getInstance("RSA");
+		} catch (NoSuchAlgorithmException ex) {
+
+			throw new ExceptionInInitializerError(ex);
+		}
+	}
+
+	public static RsaVerificationKey2018Loader get() {
+
+		return instance;
+	}
+
+	@Override
+	public RSAPublicKey verificationKey(uniresolver.did.PublicKey didPublicKey) throws GeneralSecurityException {
+
+		if (! didPublicKey.isType(DID_PUBLIC_KEY_TYPE)) {
+
+			if (log.isDebugEnabled()) log.debug("Public key " + didPublicKey.getId() + " is not of type " + DID_PUBLIC_KEY_TYPE);
+			return null;
+		}
+
+		String verifyingKeyString = didPublicKey.getPublicKeyPem();
+		if (verifyingKeyString == null) {
+
+			if (log.isDebugEnabled()) log.debug("Public key " + didPublicKey.getId() + " has no data.");
+			return null;
+		}
+
+		verifyingKeyString = verifyingKeyString.replace("-----BEGIN PUBLIC KEY-----", "");
+		verifyingKeyString = verifyingKeyString.replace("-----END PUBLIC KEY-----", "");
+		verifyingKeyString = verifyingKeyString.replace("\\r", "");
+		verifyingKeyString = verifyingKeyString.replace("\\n", "");
+
+		X509EncodedKeySpec keySpec = new X509EncodedKeySpec(Base64.decodeBase64(verifyingKeyString));
+
+		return (RSAPublicKey) keyFactory.generatePublic(keySpec);
+
+	}
+}

--- a/src/main/java/ca/bcgov/didauth/rp/verificationkeys/Secp256k1VerificationKey2018Loader.java
+++ b/src/main/java/ca/bcgov/didauth/rp/verificationkeys/Secp256k1VerificationKey2018Loader.java
@@ -1,0 +1,45 @@
+package ca.bcgov.didauth.rp.verificationkeys;
+
+import java.security.GeneralSecurityException;
+
+import org.apache.commons.codec.binary.Base64;
+import org.bitcoinj.core.Base58;
+import org.bitcoinj.core.ECKey;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.spongycastle.util.encoders.Hex;
+
+public class Secp256k1VerificationKey2018Loader implements VerificationKeyLoader<ECKey> {
+
+	private static final String DID_PUBLIC_KEY_TYPE = "Secp256k1VerificationKey2018";
+
+	private static Logger log = LoggerFactory.getLogger(Secp256k1VerificationKey2018Loader.class);
+	private static final Secp256k1VerificationKey2018Loader instance = new Secp256k1VerificationKey2018Loader();
+
+	public static Secp256k1VerificationKey2018Loader get() {
+
+		return instance;
+	}
+
+	@Override
+	public ECKey verificationKey(uniresolver.did.PublicKey didPublicKey) throws GeneralSecurityException {
+
+		if (! didPublicKey.isType(DID_PUBLIC_KEY_TYPE)) {
+
+			if (log.isDebugEnabled()) log.debug("Public key " + didPublicKey.getId() + " is not of type " + DID_PUBLIC_KEY_TYPE);
+			return null;
+		}
+
+		String verificationKeyString;
+
+		verificationKeyString = didPublicKey.getPublicKeyBase64();
+		if (verificationKeyString != null) return ECKey.fromPublicOnly(Base64.decodeBase64(verificationKeyString));
+		verificationKeyString = didPublicKey.getPublicKeyBase58();
+		if (verificationKeyString != null) return ECKey.fromPublicOnly(Base58.decode(verificationKeyString));
+		verificationKeyString = didPublicKey.getPublicKeyHex();
+		if (verificationKeyString != null) return ECKey.fromPublicOnly(Hex.decode(verificationKeyString));
+
+		if (log.isDebugEnabled()) log.debug("Public key " + didPublicKey.getId() + " has no data.");
+		return null;
+	}
+}

--- a/src/main/java/ca/bcgov/didauth/rp/verificationkeys/VerificationKeyLoader.java
+++ b/src/main/java/ca/bcgov/didauth/rp/verificationkeys/VerificationKeyLoader.java
@@ -1,0 +1,8 @@
+package ca.bcgov.didauth.rp.verificationkeys;
+
+import java.security.GeneralSecurityException;
+
+public interface VerificationKeyLoader<KEY> {
+
+	public abstract KEY verificationKey(uniresolver.did.PublicKey didPublicKey) throws GeneralSecurityException;
+}

--- a/src/main/webapp/index.jsp
+++ b/src/main/webapp/index.jsp
@@ -45,7 +45,7 @@
 		<h1>DID Auth Example</h1>
 		<p>This is an example DID Auth relying party.<p>
 		<p>It requires the DID Auth browser plugin to be installed in Firefox:<br>
-		<a href="https://github.com/peacekeeper/did-auth-relying-party">https://github.com/peacekeeper/did-auth-extension</a></p>
+		<a href="https://github.com/bcgov/did-auth-relying-party">https://github.com/bcgov/did-auth-extension</a></p>
 
 		<img src="did-auth-icon.png" align="middle">
 		<button onclick="onButtonClick()">DID Auth</button>

